src/nbl/video/IPhysicalDevice.cpp

```diff
@@ -376,125 +376,125 @@ asset::E_FORMAT narrowDownFormatPromotion(const core::unordered_set<asset::E_FOR
+//         getImageAspects(req.originalFormat).hasFlags(asset::IImage::EAF_COLOR_BIT)
+//     );
+//     auto& buf_cache = this->m_formatPromotionCache.buffers;
+//     auto cached = buf_cache.find(req);
+//     if (cached != buf_cache.end())
+//         return cached->second;
+
+//     if (req.usages < getBufferFormatUsages()[req.originalFormat])
+//     {
+//         buf_cache.insert(cached, { req,req.originalFormat });
+//         return req.originalFormat;
+//     }
+
+//     auto srcFormat = req.originalFormat;
+//     bool srcIntFormat = asset::isIntegerFormat(srcFormat);
+//     bool srcSignedFormat = asset::isSignedFormat(srcFormat);
+//     auto srcChannels = asset::getFormatChannelCount(srcFormat);
+//     double srcMinVal[4];
+//     double srcMaxVal[4];
+//     for (uint32_t channel = 0; channel < srcChannels; channel++)
+//     {
+//         srcMinVal[channel] = asset::getFormatMinValue<double>(srcFormat, channel);
+//         srcMaxVal[channel] = asset::getFormatMaxValue<double>(srcFormat, channel);
+//     }
+
+//     // Cache valid formats per usage?
+//     core::unordered_set<asset::E_FORMAT> validFormats;
+
+//     for (uint32_t format = 0; format < asset::E_FORMAT::EF_UNKNOWN; format++)
+//     {
+//         auto f = static_cast<asset::E_FORMAT>(format);
+//         // Can't have less aspects
+//         if (!getImageAspects(f).hasFlags(asset::IImage::EAF_COLOR_BIT))
+//             continue;
+
+//         if (!canPromoteFormat(f, srcFormat, srcSignedFormat, srcIntFormat, srcChannels, srcMinVal, srcMaxVal))
+//             continue;
+
+//         if (req.usages < getBufferFormatUsages()[f])
+//         {
+//             validFormats.insert(f);
+//         }
+//     }
+
+//     auto promoted = narrowDownFormatPromotion(validFormats, req.originalFormat);
+//     buf_cache.insert(cached, { req,promoted });
+//     return promoted;
+// }
+
+// asset::E_FORMAT IPhysicalDevice::promoteImageFormat(const SImageFormatPromotionRequest req, const IGPUImage::E_TILING tiling)
+// {
+//     format_image_cache_t& cache = tiling == IGPUImage::E_TILING::ET_LINEAR 
+//         ? this->m_formatPromotionCache.linearTilingImages 
+//         : this->m_formatPromotionCache.optimalTilingImages;
+//     auto cached = cache.find(req);
+//     if (cached != cache.end())
+//         return cached->second;
+
+//     auto getImageFormatUsagesTiling = [&](asset::E_FORMAT f)
+//     {
+//         switch (tiling)
+//         {
+//             case IGPUImage::E_TILING::ET_LINEAR:
+//                 return getImageFormatUsagesLinearTiling()[f];
+//             case IGPUImage::E_TILING::ET_OPTIMAL:
+//                 return getImageFormatUsagesOptimalTiling()[f];
+//             default:
+//                 assert(false); // Invalid tiling
+//         }
+//         return SFormatImageUsages::SUsage{}; // compiler please shut up
+//     };
+
+//     if (req.usages < getImageFormatUsagesTiling(req.originalFormat))
```

Comment:
clang complains about this scalar comparison

```
[build] FAILED: src/nbl/CMakeFiles/Nabla.dir/video/IPhysicalDevice.cpp.o 
[build] /usr/bin/clang++-15 -DPNG_NO_MMX_CODE -DPNG_NO_MNG_FEATURES -DPNG_THREAD_UNSAFE_OK -DSPIRV_CROSS_EXCEPTIONS_TO_ASSERTIONS -D_7ZIP_ST -D__NBL_BUILDING_NABLA__ -I/home/michaelpollind/projects/Nabla/3rdparty/parallel-hashmap/parallel_hashmap -I/home/michaelpollind/projects/Nabla/3rdparty/tcpp -I/home/michaelpollind/projects/Nabla/build/3rdparty/zlib/copy_source -I/home/michaelpollind/projects/Nabla/3rdparty/shaderc/libshaderc/include -I/home/michaelpollind/projects/Nabla/src/nbl/BEFORE -I/home/michaelpollind/projects/Nabla/3rdparty/SPIRV-Tools/include -I/home/michaelpollind/projects/Nabla/3rdparty/simdjson -I/home/michaelpollind/projects/Nabla/3rdparty/libjpeg-turbo -I/home/michaelpollind/projects/Nabla/3rdparty/libpng -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/OpenEXR/IlmImf -I/home/michaelpollind/projects/Nabla/src/nbl/Nabla -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/IlmBase/Half -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/IlmBase/Iex -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/IlmBase/IexMath -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/IlmBase/Imath -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/IlmBase/IlmThread -I/home/michaelpollind/projects/Nabla/3rdparty/gli -I/home/michaelpollind/projects/Nabla/src/nbl/X11_Xrandr_INCLUDE_PATH -I/home/michaelpollind/projects/Nabla/3rdparty/Vulkan-Headers/include -I/home/michaelpollind/projects/Nabla/3rdparty/volk -I/home/michaelpollind/projects/Nabla/build/src/nbl/include -I/home/michaelpollind/projects/Nabla/include -I/home/michaelpollind/projects/Nabla/3rdparty/glm -I/home/michaelpollind/projects/Nabla/3rdparty/renderdoc -I/home/michaelpollind/projects/Nabla/build/3rdparty/zlib -I/home/michaelpollind/projects/Nabla/build/3rdparty/openexr/IlmBase/config -I/home/michaelpollind/projects/Nabla/build/3rdparty/openexr/OpenEXR/config -I/home/michaelpollind/projects/Nabla/build/3rdparty/libpng -I/home/michaelpollind/projects/Nabla/build/3rdparty/libjpeg-turbo -I/home/michaelpollind/projects/Nabla/3rdparty -I/home/michaelpollind/projects/Nabla/build/include/nbl/config/debug -I/home/michaelpollind/projects/Nabla/src -I/home/michaelpollind/projects/Nabla/source/Nabla -I/home/michaelpollind/projects/Nabla/src/3rdparty -I/home/michaelpollind/projects/Nabla/build/include -msse4.1 -g -std=c++20 -Winvalid-pch -Xclang -include-pch -Xclang /home/michaelpollind/projects/Nabla/build/src/nbl/CMakeFiles/Nabla.dir/cmake_pch.hxx.pch -Xclang -include -Xclang /home/michaelpollind/projects/Nabla/build/src/nbl/CMakeFiles/Nabla.dir/cmake_pch.hxx -MD -MT src/nbl/CMakeFiles/Nabla.dir/video/IPhysicalDevice.cpp.o -MF src/nbl/CMakeFiles/Nabla.dir/video/IPhysicalDevice.cpp.o.d -o src/nbl/CMakeFiles/Nabla.dir/video/IPhysicalDevice.cpp.o -c /home/michaelpollind/projects/Nabla/src/nbl/video/IPhysicalDevice.cpp
[build] /home/michaelpollind/projects/Nabla/src/nbl/video/IPhysicalDevice.cpp:108:13: warning: 136 enumeration values not handled in switch: 'EF_D16_UNORM', 'EF_X8_D24_UNORM_PACK32', 'EF_D32_SFLOAT'... [-Wswitch]
[build]     switch (format)
[build]             ^~~~~~
[build] /home/michaelpollind/projects/Nabla/src/nbl/video/IPhysicalDevice.cpp:456:9: error: cannot compile this scalar expression yet
[build]     if (req.usages < getImageFormatUsagesTiling(req.originalFormat))
[build]         ^~~~~~~~~~
[build] PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
[build] Stack dump:
[build] 0.	Program arguments: /usr/bin/clang++-15 -DPNG_NO_MMX_CODE -DPNG_NO_MNG_FEATURES -DPNG_THREAD_UNSAFE_OK -DSPIRV_CROSS_EXCEPTIONS_TO_ASSERTIONS -D_7ZIP_ST -D__NBL_BUILDING_NABLA__ -I/home/michaelpollind/projects/Nabla/3rdparty/parallel-hashmap/parallel_hashmap -I/home/michaelpollind/projects/Nabla/3rdparty/tcpp -I/home/michaelpollind/projects/Nabla/build/3rdparty/zlib/copy_source -I/home/michaelpollind/projects/Nabla/3rdparty/shaderc/libshaderc/include -I/home/michaelpollind/projects/Nabla/src/nbl/BEFORE -I/home/michaelpollind/projects/Nabla/3rdparty/SPIRV-Tools/include -I/home/michaelpollind/projects/Nabla/3rdparty/simdjson -I/home/michaelpollind/projects/Nabla/3rdparty/libjpeg-turbo -I/home/michaelpollind/projects/Nabla/3rdparty/libpng -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/OpenEXR/IlmImf -I/home/michaelpollind/projects/Nabla/src/nbl/Nabla -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/IlmBase/Half -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/IlmBase/Iex -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/IlmBase/IexMath -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/IlmBase/Imath -I/home/michaelpollind/projects/Nabla/3rdparty/openexr/IlmBase/IlmThread -I/home/michaelpollind/projects/Nabla/3rdparty/gli -I/home/michaelpollind/projects/Nabla/src/nbl/X11_Xrandr_INCLUDE_PATH -I/home/michaelpollind/projects/Nabla/3rdparty/Vulkan-Headers/include -I/home/michaelpollind/projects/Nabla/3rdparty/volk -I/home/michaelpollind/projects/Nabla/build/src/nbl/include -I/home/michaelpollind/projects/Nabla/include -I/home/michaelpollind/projects/Nabla/3rdparty/glm -I/home/michaelpollind/projects/Nabla/3rdparty/renderdoc -I/home/michaelpollind/projects/Nabla/build/3rdparty/zlib -I/home/michaelpollind/projects/Nabla/build/3rdparty/openexr/IlmBase/config -I/home/michaelpollind/projects/Nabla/build/3rdparty/openexr/OpenEXR/config -I/home/michaelpollind/projects/Nabla/build/3rdparty/libpng -I/home/michaelpollind/projects/Nabla/build/3rdparty/libjpeg-turbo -I/home/michaelpollind/projects/Nabla/3rdparty -I/home/michaelpollind/projects/Nabla/build/include/nbl/config/debug -I/home/michaelpollind/projects/Nabla/src -I/home/michaelpollind/projects/Nabla/source/Nabla -I/home/michaelpollind/projects/Nabla/src/3rdparty -I/home/michaelpollind/projects/Nabla/build/include -msse4.1 -g -std=c++20 -Winvalid-pch -Xclang -include-pch -Xclang /home/michaelpollind/projects/Nabla/build/src/nbl/CMakeFiles/Nabla.dir/cmake_pch.hxx.pch -Xclang -include -Xclang /home/michaelpollind/projects/Nabla/build/src/nbl/CMakeFiles/Nabla.dir/cmake_pch.hxx -MD -MT src/nbl/CMakeFiles/Nabla.dir/video/IPhysicalDevice.cpp.o -MF src/nbl/CMakeFiles/Nabla.dir/video/IPhysicalDevice.cpp.o.d -o src/nbl/CMakeFiles/Nabla.dir/video/IPhysicalDevice.cpp.o -c /home/michaelpollind/projects/Nabla/src/nbl/video/IPhysicalDevice.cpp
[build] 1.	<eof> parser at end of file
[build] 2.	/home/michaelpollind/projects/Nabla/src/nbl/video/IPhysicalDevice.cpp:3:11: LLVM IR generation of declaration 'nbl'
[build] 3.	/home/michaelpollind/projects/Nabla/src/nbl/video/IPhysicalDevice.cpp:433:34: Generating code for declaration 'nbl::video::IPhysicalDevice::promoteImageFormat'
[build] Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
[build] /usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamEi+0x31)[0x7f61fd307281]
[build] /usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1(_ZN4llvm3sys17RunSignalHandlersEv+0xee)[0x7f61fd304f9e]
[build] /usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1(_ZN4llvm3sys15CleanupOnSignalEm+0x101)[0x7f61fd306641]
[build] /usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1(+0xe2954f)[0x7f61fd22954f]
[build] /lib/x86_64-linux-gnu/libc.so.6(+0x3bcf0)[0x7f61fbc3bcf0]
[build] /usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1(_ZN4llvm10UndefValue3getEPNS_4TypeE+0x1e)[0x7f61fd378d6e]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(+0x1b1bf35)[0x7f620511bf35]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(+0x1b28124)[0x7f6205128124]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(+0x1b1b16c)[0x7f620511b16c]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction14EmitScalarExprEPKNS_4ExprEb+0x5d)[0x7f620511391d]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction18EvaluateExprAsBoolEPKNS_4ExprE+0x19a)[0x7f62050c146a]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction20EmitBranchOnBoolExprEPKNS_4ExprEPN4llvm10BasicBlockES7_mNS_4Stmt10LikelihoodE+0x20d)[0x7f620529064d]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction10EmitIfStmtERKNS_6IfStmtE+0x4e5)[0x7f6205228d85]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction8EmitStmtEPKNS_4StmtEN4llvm8ArrayRefIPKNS_4AttrEEE+0x305)[0x7f6205227d55]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction28EmitCompoundStmtWithoutScopeERKNS_12CompoundStmtEbNS0_12AggValueSlotE+0x200)[0x7f6205233e90]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction16EmitFunctionBodyEPKNS_4StmtE+0x123)[0x7f620528ea83]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction12GenerateCodeENS_10GlobalDeclEPN4llvm8FunctionERKNS0_14CGFunctionInfoE+0x635)[0x7f620528f6e5]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule28EmitGlobalFunctionDefinitionENS_10GlobalDeclEPN4llvm11GlobalValueE+0x111)[0x7f62052ada71]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule20EmitGlobalDefinitionENS_10GlobalDeclEPN4llvm11GlobalValueE+0x24b)[0x7f62052a72fb]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule10EmitGlobalENS_10GlobalDeclE+0x3e5)[0x7f62052ab025]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule16EmitTopLevelDeclEPNS_4DeclE+0x91d)[0x7f62052a697d]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule15EmitDeclContextEPKNS_11DeclContextE+0x2b)[0x7f62052b3deb]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule16EmitTopLevelDeclEPNS_4DeclE+0x1d3)[0x7f62052a6233]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule15EmitDeclContextEPKNS_11DeclContextE+0x2b)[0x7f62052b3deb]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule16EmitTopLevelDeclEPNS_4DeclE+0x1d3)[0x7f62052a6233]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(+0x1d4e20f)[0x7f620534e20f]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(+0x1c84f30)[0x7f6205284f30]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang8ParseASTERNS_4SemaEbb+0x309)[0x7f620408e849]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang13CodeGenAction13ExecuteActionEv+0xd5)[0x7f6205281785]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang14FrontendAction7ExecuteEv+0x67)[0x7f6205cbe4b7]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang16CompilerInstance13ExecuteActionERNS_14FrontendActionE+0x336)[0x7f6205c2f606]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang25ExecuteCompilerInvocationEPNS_16CompilerInstanceE+0x29a)[0x7f6205d3ca6a]
[build] /usr/bin/clang++-15(_Z8cc1_mainN4llvm8ArrayRefIPKcEES2_Pv+0x968)[0x55fae6f2e598]
[build] /usr/bin/clang++-15(+0x1169b)[0x55fae6f2c69b]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(+0x228fbd2)[0x7f620588fbd2]
[build] /usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1(_ZN4llvm20CrashRecoveryContext9RunSafelyENS_12function_refIFvvEEE+0xde)[0x7f61fd2292ae]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZNK5clang6driver10CC1Command7ExecuteEN4llvm8ArrayRefINS2_8OptionalINS2_9StringRefEEEEEPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPb+0x138)[0x7f620588f6a8]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZNK5clang6driver11Compilation14ExecuteCommandERKNS0_7CommandERPS3_b+0x44a)[0x7f620585171a]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZNK5clang6driver11Compilation11ExecuteJobsERKNS0_7JobListERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEEb+0x8e)[0x7f620585196e]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang6driver6Driver18ExecuteCompilationERNS0_11CompilationERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEE+0x39f)[0x7f620586e99f]
[build] /usr/bin/clang++-15(_Z10clang_mainiPPc+0x27ef)[0x55fae6f2be1f]
[build] /lib/x86_64-linux-gnu/libc.so.6(+0x23510)[0x7f61fbc23510]
[build] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x89)[0x7f61fbc235c9]
[build] /usr/bin/clang++-15(_start+0x25)[0x55fae6f292e5]
```

Reply:
you probably have to explicitly name the return type on the `getImageFormatUsagesTiling lambda

Also what's going on with that `format` switch @Erfan-Ahmadi ?

Reply:
so I cut everything out of the function and found that i can still break clang 

```
auto srcFormat = req.originalFormat;
```

```
build] 1.	<eof> parser at end of file
[build] 2.	/home/michaelpollind/projects/Nabla/src/nbl/video/IPhysicalDevice.cpp:3:11: LLVM IR generation of declaration 'nbl'
[build] 3.	/home/michaelpollind/projects/Nabla/src/nbl/video/IPhysicalDevice.cpp:433:34: Generating code for declaration 'nbl::video::IPhysicalDevice::promoteImageFormat'
[build] Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
[build] /usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1(_ZN4llvm3sys15PrintStackTraceERNS_11raw_ostreamEi+0x31)[0x7f03c0707281]
[build] /usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1(_ZN4llvm3sys17RunSignalHandlersEv+0xee)[0x7f03c0704f9e]
[build] /usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1(_ZN4llvm3sys15CleanupOnSignalEm+0x101)[0x7f03c0706641]
[build] /usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1(+0xe2954f)[0x7f03c062954f]
[build] /lib/x86_64-linux-gnu/libc.so.6(+0x3bcf0)[0x7f03bf03bcf0]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZNK5clang10ASTContext15getTypeInfoImplEPKNS_4TypeE+0x610)[0x7f03c7574cc0]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZNK5clang10ASTContext11getTypeInfoEPKNS_4TypeE+0xd5)[0x7f03c7575c55]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZNK5clang10ASTContext21getPreferredTypeAlignEPKNS_4TypeE+0x32)[0x7f03c7575512]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZNK5clang10ASTContext12getDeclAlignEPKNS_4DeclEb+0x480)[0x7f03c7574280]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction17EmitAutoVarAllocaERKNS_7VarDeclE+0x75)[0x7f03c84a7525]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction11EmitVarDeclERKNS_7VarDeclE+0xd3)[0x7f03c84a42c3]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction8EmitDeclERKNS_4DeclE+0xe5)[0x7f03c84a3f85]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction14EmitSimpleStmtEPKNS_4StmtEN4llvm8ArrayRefIPKNS_4AttrEEE+0x4cc)[0x7f03c862867c]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction8EmitStmtEPKNS_4StmtEN4llvm8ArrayRefIPKNS_4AttrEEE+0xc1)[0x7f03c8627b11]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction28EmitCompoundStmtWithoutScopeERKNS_12CompoundStmtEbNS0_12AggValueSlotE+0x200)[0x7f03c8633e90]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction16EmitFunctionBodyEPKNS_4StmtE+0x123)[0x7f03c868ea83]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen15CodeGenFunction12GenerateCodeENS_10GlobalDeclEPN4llvm8FunctionERKNS0_14CGFunctionInfoE+0x635)[0x7f03c868f6e5]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule28EmitGlobalFunctionDefinitionENS_10GlobalDeclEPN4llvm11GlobalValueE+0x111)[0x7f03c86ada71]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule20EmitGlobalDefinitionENS_10GlobalDeclEPN4llvm11GlobalValueE+0x24b)[0x7f03c86a72fb]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule10EmitGlobalENS_10GlobalDeclE+0x3e5)[0x7f03c86ab025]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule16EmitTopLevelDeclEPNS_4DeclE+0x91d)[0x7f03c86a697d]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule15EmitDeclContextEPKNS_11DeclContextE+0x2b)[0x7f03c86b3deb]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule16EmitTopLevelDeclEPNS_4DeclE+0x1d3)[0x7f03c86a6233]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule15EmitDeclContextEPKNS_11DeclContextE+0x2b)[0x7f03c86b3deb]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang7CodeGen13CodeGenModule16EmitTopLevelDeclEPNS_4DeclE+0x1d3)[0x7f03c86a6233]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(+0x1d4e20f)[0x7f03c874e20f]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(+0x1c84f30)[0x7f03c8684f30]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang8ParseASTERNS_4SemaEbb+0x309)[0x7f03c748e849]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang13CodeGenAction13ExecuteActionEv+0xd5)[0x7f03c8681785]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang14FrontendAction7ExecuteEv+0x67)[0x7f03c90be4b7]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang16CompilerInstance13ExecuteActionERNS_14FrontendActionE+0x336)[0x7f03c902f606]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang25ExecuteCompilerInvocationEPNS_16CompilerInstanceE+0x29a)[0x7f03c913ca6a]
[build] /usr/bin/clang++-15(_Z8cc1_mainN4llvm8ArrayRefIPKcEES2_Pv+0x968)[0x55ebe2c62598]
[build] /usr/bin/clang++-15(+0x1169b)[0x55ebe2c6069b]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(+0x228fbd2)[0x7f03c8c8fbd2]
[build] /usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1(_ZN4llvm20CrashRecoveryContext9RunSafelyENS_12function_refIFvvEEE+0xde)[0x7f03c06292ae]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZNK5clang6driver10CC1Command7ExecuteEN4llvm8ArrayRefINS2_8OptionalINS2_9StringRefEEEEEPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPb+0x138)[0x7f03c8c8f6a8]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZNK5clang6driver11Compilation14ExecuteCommandERKNS0_7CommandERPS3_b+0x44a)[0x7f03c8c5171a]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZNK5clang6driver11Compilation11ExecuteJobsERKNS0_7JobListERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEEb+0x8e)[0x7f03c8c5196e]
[build] /usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15(_ZN5clang6driver6Driver18ExecuteCompilationERNS0_11CompilationERN4llvm15SmallVectorImplISt4pairIiPKNS0_7CommandEEEE+0x39f)[0x7f03c8c6e99f]
[build] /usr/bin/clang++-15(_Z10clang_mainiPPc+0x27ef)[0x55ebe2c5fe1f]
[build] /lib/x86_64-linux-gnu/libc.so.6(+0x23510)[0x7f03bf023510]
[build] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x89)[0x7f03bf0235c9]
[build] /usr/bin/clang++-15(_start+0x25)[0x55ebe2c5d2e5]
[build] clang: error: clang frontend command failed with exit code 139 (use -v to see invocation)
[build] Ubuntu clang version 15.0.6
[build] Target: x86_64-pc-linux-gnu
[build] Thread model: posix
[build] InstalledDir: /usr/bin
[build] clang: note: diagnostic msg: 
[build] ********************
[build] 
[build] PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
[build] Preprocessed source(s) and associated run script(s) are located at:
[build] clang: note: diagnostic msg: /tmp/IPhysicalDevice-cb4ead.cpp
[build] clang: note: diagnostic msg: /tmp/IPhysicalDevice-cb4ead.sh
[build] clang: note: diagnostic msg: 
[build] 
[build] ********************
```

Reply:
> you probably have to explicitly name the return type on the `getImageFormatUsagesTiling lambda
> 
> Also what's going on with that `format` switch @Erfan-Ahmadi ?

The format switch case does not have a default. It's asking about the other cases that are not included.

Reply:
where is even the original switch?

Reply:
This is @deprilula28 's domain. Please let him know about your inquiries.

Reply:
> where is even the original switch?

That's a minor warning that is something else that can be looked into. There are a lot of other clang warnings. We can look into later.